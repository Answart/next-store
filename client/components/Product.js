import PropTypes from 'prop-types';
import Link from 'next/link';
import StyledProduct from './styles/ProductStyles';
import ByCreator from './ByCreator';
import ProductVariants from './ProductVariants';


const Product = props => {
  const { product } = props;
  return (
    <StyledProduct>
      <div className="prdct-imgs">
        <img width="450" height="640" src={product.image.large_image_url} alt={product.title} />
      </div>

      <div className="prdct-content">
        <div id='prdct-title' className="prdct-padding">
          <Link href={{
            pathname: '/buy',
            query: { id: product.id }
          }}>
            <a className="prdct-title">
              {product.title}
            </a>
          </Link>
        </div>

        <ByCreator
          name={product.user.name}
          online={true}
        />

        <div id='prdct-online' className="prdct-padding">
          {!product.online && (
            <i>Offline</i>
          )}
        </div>

        <ProductVariants
          variants={product.variants}
          online={product.online}
          demoView={props.demoView}
          VariantActionComponent={props.variantActionComponent}
          variantAction={props.variantAction}
          variantActionLabel={props.variantActionLabel}
        />

        <div id='prdct-desc' className="prdct-desc prdct-padding">
          <div id="prdct-description">
            <strong>
              Description:
            </strong>

            <p>
              {product.description}
            </p>
          </div>

          {product.brand && (
            <div id="prdct-brand">
              <strong>
                Brand:
              </strong>

              <p>
                {product.brand}
              </p>
            </div>
          )}
        </div>
      </div>
    </StyledProduct>
  );
};

Product.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    department: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    category: PropTypes.string,
    brand: PropTypes.string,
    online: PropTypes.bool.isRequired,
    image: PropTypes.shape({
      id: PropTypes.string.isRequired,
      cloudinary_id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      width: PropTypes.number.isRequired,
      height: PropTypes.number.isRequired,
      transformation: PropTypes.string.isRequired,
      image_url: PropTypes.string.isRequired,
      large_image_url: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  demoView: PropTypes.bool,
  VariantActionComponent: PropTypes.func,
  variantAction: PropTypes.func,
  variantActionLabel: PropTypes.string
};


export default Product;
