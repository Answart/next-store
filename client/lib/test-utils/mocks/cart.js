import {
  ADD_TO_CART_MUTATION,
  UPDATE_CARTITEM_MUTATION,
  REMOVE_FROM_CART_MUTATION
} from '../../../graphql'
import { fakeCartItem } from '../utils';
import { mockVariant } from './variant';

const mockCartItem = fakeCartItem();


const addToCartMutationMock = {
  request: { query: ADD_TO_CART_MUTATION, variables: { id: mockVariant.id } },
  result: {
    data: {
      addToCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};

const updateCartItemMutationMock = (quantity) => ({
  request: { query: UPDATE_CARTITEM_MUTATION, variables: { id: mockCartItem.id, quantity } },
  result: {
    data: {
      updateCartItem: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
        quantity,
        variant: { ...mockCartItem.variant }
      },
    },
  },
});

const removeFromCartMutationMock = {
  request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: mockCartItem.id } },
  result: {
    data: {
      removeFromCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};


export {
  mockCartItem,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
}
