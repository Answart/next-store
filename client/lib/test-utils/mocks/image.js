import { CREATE_IMAGE_MUTATION } from '../../../graphql'
import { fakeImage } from '../utils';

const mockImage = fakeImage();
const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
}


const createImageMutationMock = (overrides) => ({
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      ...overrides,
    },
  },
  result: {
    data: {
      createImage: {
        __typename: 'Image',
        id: mockImage.id,
      },
    },
  },
});

const createImageMutationErrorMock = {
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      cloudinary_id: '',
      name: '',
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};


export {
  mockImage, mockImageVariables,
  createImageMutationMock, createImageMutationErrorMock,
}
