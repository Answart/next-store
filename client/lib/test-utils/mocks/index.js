export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
