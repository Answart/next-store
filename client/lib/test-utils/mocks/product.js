import {
  PRODUCT_QUERY, SHOP_PRODUCTS_QUERY,
  CREATE_PRODUCT_MUTATION,
  UPDATE_PRODUCT_MUTATION,
  DELETE_PRODUCT_MUTATION,
} from '../../../graphql'
import { mockUser, mockProduct, mockImage, mockVariant, mockShopProductsVariables } from '../utils';


const productQueryMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage
          }
        }]
      },
    },
  },
};

const productQueryNoVariantMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: []
      },
    },
  },
};

const createProductMutationMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      createProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const createProductMutationErrorMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const updateProductMutationMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      updateProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const updateProductMutationErrorMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const deleteProductMutationMock = {
  request: { query: DELETE_PRODUCT_MUTATION, variables: { id: mockProduct.id } },
  result: {
    data: {
      deleteProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const shopProductsQueryNameEmptyMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { name: mockUser.name }},
  result: {
    data: {
      products: [],
    },
  },
};

const shopProductsQueryProductMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { ...mockShopProductsVariables }},
  result: {
    data: {
      products: [{
        ...mockProduct,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage,
          }
        }]
      }],
    },
  },
};


export {
  productQueryMock, productQueryNoVariantMock,
  shopProductsQueryNameEmptyMock, shopProductsQueryProductMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
}
