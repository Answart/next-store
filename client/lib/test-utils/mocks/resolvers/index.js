export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
} from './local';
export {
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
} from './user';
export {
  createImageMutationMock, createImageMutationErrorMock,
} from './image';
export {
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
} from './product';
export {
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock
} from './variant';
export {
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
} from './cartItem';
