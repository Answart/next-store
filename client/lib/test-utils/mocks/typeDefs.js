import {
  fakeUser, fakeImage, fakeProduct, fakeVariant, fakeCartItem, fakeOrder, fakeOrderItem,
} from '../utils';


const mockUser = fakeUser();
const mockUser2 = fakeUser();
const mockImage = fakeImage();
const mockProduct = fakeProduct();
const mockVariant = fakeVariant();
const mockCartItem = fakeCartItem();
const mockOrder = fakeOrder();
const mockOrderItem = fakeOrderItem();

mockOrder.items.push(mockOrderItem);

const mockUsers = [
  mockUser,
  {
    ...mockUser2,
    id: '87654321',
    permissions: ['USER'],
  }
];
const mockProducts = [{
  ...mockProduct,
  image: mockImage,
  user: {
    __typename: mockUser.__typename,
    id: mockUser.id,
    name: mockUser.name,
  },
  variants: [{
    ...mockVariant,
    product: {
      __typename: mockProduct.__typename,
      id: mockProduct.id,
      image: mockImage,
    }
  }]
}];

const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
};

const mockShopProductsVariables = {
  name: mockUser.name,
  orderBy: 'createdAt_DESC',
  skip: 0,
  first: 1
 };


export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrder, mockOrderItem,
};
