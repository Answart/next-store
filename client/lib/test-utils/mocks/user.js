import {
  CURRENT_USER_QUERY,
  SIGNOUT_MUTATION,
  REQUEST_PASSWORD_RESET_MUTATION,
} from '../../../graphql'
import { fakeUser } from '../utils';
import { mockCartItem } from './cart';

const mockUser = fakeUser();


const userQueryEmptyCartMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [],
      },
    },
  },
};

const userQueryNoUserMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: { me: null }
  },
};

const userQueryCartItemMock = overrides => ({
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [{
          ...mockCartItem,
          ...overrides
        }],
      },
    },
  },
});

const signoutMutationMock = {
  request: { query: SIGNOUT_MUTATION },
  result: {
    data: {
      signout: {
        __typename: 'Message',
        success: true,
        message: 'Goodbye!'
      },
    },
  },
};

const requestPasswordResetMutationMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'answart@sbcglobal.net' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: true,
        message: 'Your request has been sent! Check your email.',
      }
    },
  },
};

const requestPasswordResetMutationErrorMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'err@g.c' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: false,
        message: 'No such user found for email err@g.c',
      }
    },
  },
};


export {
  mockUser,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock
};