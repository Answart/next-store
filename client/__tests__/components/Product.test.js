import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import Product from '../../components/Product';
import { mockProduct } from '../../lib/test-utils/mocks';


describe('<Product />', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider mocks={[]}>
        <Product
          product={mockProduct}
          demoView={false}
          variantAction={jest.fn()}
          variantActionLabel='Select'
        />
      </MockedProvider>
    );
  });
  afterAll(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it('matches the snap shot', async () => {
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('Product'))).toMatchSnapshot();
  });

  it('renders properly', async () => {
    expect(wrapper.find('.prdct-imgs').length).toBe(1);
    expect(wrapper.find('#prdct-title').length).toBe(1);
    expect(wrapper.find('#prdct-title').length).toBe(1);
    expect(wrapper.find('#prdct-title').find('Link').props().href).toMatchObject({
      pathname: "/buy",
      query: { id: mockProduct.id }
    });
    expect(wrapper.find('#prdct-title').text()).toBe(mockProduct.title);
    expect(wrapper.find('ByCreator').length).toBe(1);
    expect(wrapper.find('ByCreator').text()).toBe(`By${mockProduct.user.name}`);
    expect(wrapper.find('#prdct-online').length).toBe(1);
    expect(wrapper.find('#prdct-online').text()).toBe('Offline');
    expect(wrapper.find('ProductVariants').length).toBe(1);
    expect(wrapper.find('#prdct-description').length).toBe(1);
    expect(wrapper.find('#prdct-description').text()).toBe(`Description:${mockProduct.description}`);
    expect(wrapper.find('#prdct-brand').length).toBe(1);
    expect(wrapper.find('#prdct-brand').text()).toBe(`Brand:${mockProduct.brand}`);
  });
});
