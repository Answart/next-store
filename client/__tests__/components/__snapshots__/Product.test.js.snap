// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> matches the snap shot 1`] = `
<Product
  demoView={false}
  product={
    Object {
      "__typename": "Product",
      "brand": "Peggs",
      "category": "jewelry",
      "department": "accessories",
      "description": "Limited Edition watch from the 2018 Fall fashion line.",
      "id": "pr0duct1d",
      "image": Object {
        "__typename": "Image",
        "cloudinary_id": "111111",
        "height": 22,
        "id": "1m4g31d",
        "image_url": "peggswatch1.jpg",
        "large_image_url": "peggswatch2.jpg",
        "name": "peggswatch1",
        "transformation": "",
        "width": 22,
      },
      "online": false,
      "title": "Peggs Gold Edition Analog Watch Peggs Gold Edition Analog ",
      "user": Object {
        "__typename": "User",
        "cart": Array [],
        "email": "Waldo.Yundt@Mosciski.biz",
        "id": "us3r1d",
        "name": "Ms. Josue McCullough",
        "password": "unique-password",
        "permissions": Array [
          "ADMIN",
        ],
      },
      "variants": Array [],
    }
  }
  variantAction={[MockFunction]}
  variantActionLabel="Select"
>
  <ProductStyles__StyledProduct>
    <div
      className="ProductStyles__StyledProduct-sc-2231u4-0 dOvnhj"
    >
      <div
        className="prdct-imgs"
      >
        <img
          alt="Peggs Gold Edition Analog Watch Peggs Gold Edition Analog "
          height="640"
          src="peggswatch2.jpg"
          width="450"
        />
      </div>
      <div
        className="prdct-content"
      >
        <div
          className="prdct-padding"
          id="prdct-title"
        >
          <Link
            href={
              Object {
                "pathname": "/buy",
                "query": Object {
                  "id": "pr0duct1d",
                },
              }
            }
          >
            <a
              className="prdct-title"
              href="/buy?id=pr0duct1d"
              onClick={[Function]}
            >
              Peggs Gold Edition Analog Watch Peggs Gold Edition Analog 
            </a>
          </Link>
        </div>
        <ByCreator
          name="Ms. Josue McCullough"
          online={true}
        >
          <div
            className="prdct-creator prdct-padding"
          >
            By
            <Link
              href={
                Object {
                  "pathname": "/shop",
                  "query": Object {
                    "name": "Ms. Josue McCullough",
                    "online": true,
                  },
                }
              }
            >
              <a
                className="undrln-btn"
                href="/shop?name=Ms.%20Josue%20McCullough&online=true"
                onClick={[Function]}
              >
                Ms. Josue McCullough
              </a>
            </Link>
          </div>
        </ByCreator>
        <div
          className="prdct-padding"
          id="prdct-online"
        >
          <i>
            Offline
          </i>
        </div>
        <ProductVariants
          demoView={false}
          online={false}
          variantAction={[MockFunction]}
          variantActionLabel="Select"
          variants={Array []}
        >
          <ProductStyles__StyledProductVariants>
            <div
              className="ProductStyles__StyledProductVariants-sc-2231u4-1 ndRtn"
            >
              <div
                className="prdct-padding"
                id="prod-var-avail"
              >
                <i>
                  Unavailable
                </i>
              </div>
            </div>
          </ProductStyles__StyledProductVariants>
        </ProductVariants>
        <div
          className="prdct-desc prdct-padding"
          id="prdct-desc"
        >
          <div
            id="prdct-description"
          >
            <strong>
              Description:
            </strong>
            <p>
              Limited Edition watch from the 2018 Fall fashion line.
            </p>
          </div>
          <div
            id="prdct-brand"
          >
            <strong>
              Brand:
            </strong>
            <p>
              Peggs
            </p>
          </div>
        </div>
      </div>
    </div>
  </ProductStyles__StyledProduct>
</Product>
`;
